pkg/storage/snapshot/source.go

```diff
@@ -179,6 +184,69 @@ func (s *vmSnapshotSource) getVMRevision() (*snapshotv1.VirtualMachine, error) {
 	return vmRevision, nil
 }
 
+func (s *vmSnapshotSource) getControllerRevision(namespace, name string) (*appsv1.ControllerRevision, error) {
+	crKey := cacheKeyFunc(namespace, name)
+	obj, exists, err := s.controller.CRInformer.GetStore().GetByKey(crKey)
+	if err != nil {
+		return nil, err
+	}
+	if !exists {
+		return nil, fmt.Errorf("Unable to fetch ControllerRevision %s", crKey)
+	}
+	return obj.(*appsv1.ControllerRevision), nil
+}
+
+func (s *vmSnapshotSource) captureInstancetypeControllerRevision(namespace, name string) (string, error) {
+	existingCR, err := s.getControllerRevision(namespace, name)
+	if err != nil {
+		return "", err
+	}
+
+	snapshotCR := existingCR.DeepCopy()
+	snapshotCR.ObjectMeta.Reset()
+
+	// We strip out the source VM name from the CR name and replace it with the snapshot name
+	snapshotCR.Name = s.snapshot.Name + strings.ReplaceAll(existingCR.Name, s.snapshot.Spec.Source.Name, "")
+
+	// Ensure GVK is set before we attempt to create the controller OwnerReference below
+	obj, err := utils.EnsureKubeVirtGroupVersionKind(s.snapshot)
+	if err != nil {
+		return "", err
+	}
+	snapshot, ok := obj.(*snapshotv1.VirtualMachineSnapshot)
+	if !ok {
+		return "", fmt.Errorf("Unexpected object format returned from EnsureKubeVirtGroupVersionKind")
+	}
+	snapshotCR.OwnerReferences = []metav1.OwnerReference{*metav1.NewControllerRef(snapshot, snapshot.GroupVersionKind())}
+
+	snapshotCR, err = s.controller.Client.AppsV1().ControllerRevisions(s.snapshot.Namespace).Create(context.Background(), snapshotCR, metav1.CreateOptions{})
+	if err != nil && !errors.IsAlreadyExists(err) {
```

Comment:
How can it happen that a CR already exists for this snapshot? Can the content of the CRs differ?

Reply:
When restoring to the original `VirtualMachine` it's actually likely to already exist, the contents shouldn't differ as the name includes the UID and generation of the resource contained within. That said we do have code within `pkg/instancetype` I could refactor and call here:

https://github.com/kubevirt/kubevirt/blob/73d78d58ca6b48c8d88ce3ec5a03e3e6355689bd/pkg/instancetype/instancetype.go#L278-L306

I'll add another commit putting this in place.

Reply:
/cc @akrejcir 

I've pushed 451a2fbaff09093d989169f1e088a4ff7923bce1 to extract and refactor the original `ControllerRevision` comparison code in the instancetypes package. I've then wired this up in e8b279702e7beb4c7021154f9f8664442436f0bd but tbh this could be a follow up PR to close this gap. Thoughts appreciated, I've marked this PR as draft for now as the above commits need some additional test coverage but the existing tests are passing locally for me.

Reply:
I've created PR #8674 to track this refactor and comparison during a restore as I'd like to land and backport this series to v0.58.0 as soon as possible. IMHO #8674 is just a useful follow up. Any objections to this @0xFelix ?

Reply:
No objections, a follow-up PR for the comparison enhancement is fine for me.